### PR TITLE
Avoid unnecessary _calc_train calls

### PR DESCRIPTION
**Pull request by axefor**
This PR suggests to more flexibly allow "run" and "train" engine calls, to avoid unnecessary "train" calls. For example, it includes handing over ``loss``, doing only E/F/S eval ("_calc_run"), as the SciPy function handle, while ``loss.jac`` goes as the jac argument. See end for some notes on performance, but in most practical cases the idea is that it improves performance.

The PR does this by reverting the "mode" based engine switch, to instead use different methods (new are ``calc.compute_jacobian()``, ``engine.efs()`` and ``engine.jac()``) to more flexibly change between ``calc_run`` and ``calc_train``.

Performance wise this is an improvement when the timing difference between ``_calc_run`` and ``_calc_train`` is large, and the number of trial evaluations is larger then the number of iterations. The trade-off is an up to x2 slower optimization when the number of iterations equals the number of trial steps, and there is no time difference between ``_calc_run`` and ``_calc_train`` as is almost the case for a level 2 MTP. This could however be a sensible trade-off since level-2 general optimization is rare and anyway quite fast. In practice, level 8 training could also be slower, depending on the optimizer choosen. Apart from this, it is of course a trivial (large) gain for gradient-free optimizers.

### CODE REVIEW

**Comment by yuzie007** [comment]
Thank you @axefor for the PR! I am basically OK to merge if it improves the performance.

I have not simply understood why we see the substantial performance change. The reason why we separated `run` and `train` modes was already to avoid unnecessary Jacobian calculations, in my understanding. The present PR probably means that my understanding was wrong, and somewhere we accidentally calculated unnecessary Jacobian. Could you point out which place we did this mistake?

Another reason of my confusion is that, `loss` is mostly for training (or, am I wrong?), and for training, in most practical cases we need Jabocian for non-linear fitting. So, I am also not yet sure when we just need E/F/S in `loss`.

(Probably I also do not yet understand `trial`:))

**Comment by axefor** [comment]
Thank you for the comment @yuzie007 !

Depending on which optimizer is used, and if for some reason one would like to run with jac=False, the basis Jacobian would not need to be computed, which allows for speedup of several times in such cases, with the current PR.

It turns out that the BFGS implementation, and several other algorithms, in SciPy, always calls both loss and "jac". Because in this case, the PR meant a pure degradation, I've now added a ``loss_and_jac`` method to ``LossFunction``, that is passed to SciPy ``minimize`` if ``jac==True``. For some implementations in ``minimize``, like the fallback python BFGS and (maybe) trust-ncg, there might be calls to only the loss function value, without the need for the Jacobian at every such call. However, they are probably not used in practice, at least at the moment, so just giving ``loss_and_jac`` might be the easiest solution now to avoid the mentioned performance degradation for SciPy BFGS.

So, while there were in practice no mistakes before, the current PR provides more flexibility for efficiently testing other optimization algorithms, where the basis Jacobian is not necessarily needed at every evaluation step. With the edit, there is now also no degradation.

**Comment by yuzie007** [comment]
Thank you @axefor for detailed explanation! OK, I agree the flexibility to compute Jacobian or not can optimize performance of various types of optimizers. It looks also more explicit to control this, not by `mode` but by the `jac` argument.

**Review by yuzie007 — APPROVED** [review]

**Comment by axefor** [comment]
PR #128 replaces this PR.
